media-libs/libyuv: add new package

N-R-K opened the pull request:
adds libyuv as a new package.
adds useflag for libyuv for imlib2 and libavif.

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [ ] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

stefson commented:
Do you think its possible to have a live ebuild of libyuv as well? This would make it very easy to scan for breakages on arm. Thank you!

N-R-K commented:
> Do you think its possible to have a live ebuild of libyuv as well? This would make it very easy to scan for breakages on arm. Thank you!

I will look into it, sure.

For now I'm a bit more concerned about ABI stability. Upstream doesn't seem to be using semver so I don't know how API/ABI stable they are.

Also, both [fedora](https://src.fedoraproject.org/rpms/libyuv/blob/rawhide/f/libyuv-0002-Use-a-proper-so-version.patch) and [debian](https://salsa.debian.org/debian/libyuv/-/blob/master/debian/patches/0001-CMakeLists.txt-Apply-SONAME-and-multiarch.patch?ref_type=heads) seem to be adding patches to do SONAME versioning. I'm considering applying it too.

For now I'll mark the PR as WIP until I'm more confident.

N-R-K commented:
I've added a live ebuild and also doing SONAME versioning similar to debian and fedora.

eli-schwartz reviewed (changes requested):
> For now I'm a bit more concerned about ABI stability. Upstream doesn't seem to be using semver so I don't know how API/ABI stable they are.
> 
> Also, both [fedora](https://src.fedoraproject.org/rpms/libyuv/blob/rawhide/f/libyuv-0002-Use-a-proper-so-version.patch) and [debian](https://salsa.debian.org/debian/libyuv/-/blob/master/debian/patches/0001-CMakeLists.txt-Apply-SONAME-and-multiarch.patch?ref_type=heads) seem to be adding patches to do SONAME versioning. I'm considering applying it too.
> 
> For now I'll mark the PR as WIP until I'm more confident.



> I've added a live ebuild and also doing SONAME versioning similar to debian and fedora.

In general adding thirdparty patches like this is more trouble than it's worth. Using SLOT= will already force packages to be rebuilt as needed. Adding the version number to the soname doesn't really make this easier, but makes the patching situation worse while also potentially clashing with any upstream decision to start using a soname -- and I notice you are STILL not doing what I said, i.e. adding SLOT= which you *need* to do whether you patch the soname or not.

N-R-K commented:
> Adding the version number to the soname doesn't really make this easier, but makes the patching situation worse while also potentially clashing with any upstream decision to start using a soname

I'm neutral on this. @parona-source suggested it's nice to have and so I applied it. Do you suggest removing it?

> adding SLOT= which you need to do whether you patch the soname or not

What do you mean by "adding SLOT"? My understanding is that you add a slot to be able to install multiple versions, which I'm not sure applies here? And you add/update sub-slot to force rebuild. But since it's a new package, there's nothing to force rebuilding?

eli-schwartz commented:
> I'm neutral on this. @parona-source suggested it's nice to have and so I applied it. Do you suggest removing it?

I strongly state this, and furthermore encourage you to research what the patch even means. But since nobody appears to have done so, I'll say it straight out.

The debian/fedora patch applies a soname of "0" and a version of "0.0.${PV}". It does nothing whatsoever for forcing rebuilds, not even with `@preserved-rebuild`. The only reason debian or fedora apply such a patch at all is because of weird distro-specific rules such as "all packages providing a shared library without a numeric soname have to be downstream patched to use a soname of zero, to indicate that they aren't ABI-stable".

As a completely separate matter, I find it questionable to patch versions into sonames at all. They are effectively redundant with SLOT= and introduce deviation from upstream sources without any actual clear benefit.

I don't really understand why @parona-source would suggest it's nice to have, either in the abstract general sense, or in the sense of this particular package without checking.

> What do you mean by "adding SLOT"? [...]

You have correctly deduced that when I said this: "Using SLOT= will already force packages to be rebuilt as needed."

I was referring to subslots.

> But since it's a new package, there's nothing to force rebuilding?

Challenge: ponder the nature of subslot rebuilds. Then ask yourself, once it is no longer a new package but has been updated from version 1904 to version 1905 (actually upstream is already at 1906), what should the value of SLOT= be?

ionenwks commented:
Just adding that I'm largely against inventing versioning as well.

N-R-K commented:
> Just adding that I'm largely against inventing versioning as well.

Okay, I'll remove it.

> You have correctly deduced that when I said this: "Using SLOT= will already force packages to be rebuilt as needed."
> 
> I was referring to subslots.
> 
> > But since it's a new package, there's nothing to force rebuilding?
> 
> Challenge: ponder the nature of subslot rebuilds. Then ask yourself, once it is no longer a new package but has been updated from version 1904 to version 1905 (actually upstream is already at 1906), what should the value of SLOT= be?

I don't really get what you want me to do here. Throwing riddles at outside contributors who aren't necessarily experienced in gentoo package maintenance is not very helpful.

eli-schwartz commented:
> I don't really get what you want me to do here. Throwing riddles at outside contributors who aren't necessarily experienced in gentoo package maintenance is not very helpful.

I can answer questions, but I also want people to think and learn from the experience. :)

When you update the version and you are concerned that ABI isn't stable, you have two choices:
- you're going to switch from `SLOT="0"` over to `SLOT="0/1905"`
- you're going to laboriously analyze the ABI surface, possibly with the benefit of tools like libabigail but remembering that those aren't perfect, and please do remember to compare the before/after using debug info, and bump it "only when needed" to `SLOT="0/1"` then on a subsequent update to `SLOT="0/2"`

Most likely you are going to do option 1. It is conventional in such cases to avoid the risk of forgetting, and use `SLOT="0/${PV}"` given that you already intend to provide a guaranteed contract that it shall be rebuilt on every update as a proactive measure.

If you did do option 2, then that is effectively reinventing ABI stability downstream (and you could then argue it makes sense to patch the CMakeLists.txt to add that same `abidiff`ed research discovery of "number of times the ABI actually changed" as a soname version). It is conventional for packages that provide libraries which do have a soname version, to set `SLOT="0/${LIBRARY_SONAME}"` for consistency and so that people can reliably tell when to bump it -- that is, when the soname changes.

Treating subslot as an incrementing counter of "the number of times I've packaged an updated ebuild" is best avoided if you can handle it "automatically".

N-R-K commented:
@eli-schwartz  Okay, thank you. That makes sense. But I think there's a small disconnect here.

When I said I'm 'concerned about ABI stability' I didn't mean that they have a history of it. In fact I tried to look it up on some distro's bug trackers and found nothing.

But rather, I get the impression (which maybe entirely incorrect) that the library was meant to be "in-house" primary and used by other secondary. And AFAIK google rebuilds everything from scratch everytime, so it *wouldn't be surprising* if ABI wasn't very stable.

But maybe that's not true since upstream does define their own MAJOR and MINOR version in `CM_linux_packages.cmake`. But are those maintained by upstream as a ABI (or at least API) stability guarantee? Or just a token effort to comply with some distro's weird versioning requirements? Going through the git log [doesn't reveal much](https://chromium.googlesource.com/libyuv/libyuv/+/33f52bdac996966f82bd870a0887ddc33b918aba%5E%21/#F0).

Going with `SLOT="0/${PV}"` would be safe. But if upstream is ABI stable then it would also cause unnecessary rebuilds, which isn't super nice. I suppose I'll just go with the safer option though, since currently all 3 packages (including libyuv itself) are not super heavy on compile times ¯\\\_(ツ)\_/¯

N-R-K commented:
> Going through the git log [doesn't reveal much](https://chromium.googlesource.com/libyuv/libyuv/+/33f52bdac996966f82bd870a0887ddc33b918aba%5E%21/#F0).

Noticed there's a [bug linked](https://libyuv.issues.chromium.org/issues/42280626#comment1) in the commit message. It still doesn't mention why they prepended `0.0.` randomly. Unless someone has evidence suggesting otherwise, I'm going to assume those MAJOR, MINOR version they're defining to be basically useless.

- - -

Changes:

- Removed the SONAME patch
- Added `SLOT="0/${PV}"`
- Added `:=` to imlib2 and libavif dependency

Curiously though, if I try to upgrade from libyuv-1904 to 9999 (after rebuilding both imlib2 and libavif against 1904) emerge triggers a rebuild for imlib2 but not libavif? Did I mess something up here?

N-R-K commented on media-libs/libyuv/libyuv-9999.ebuild:
```diff
@@ -0,0 +1,50 @@
+	cmake_src_install
+
+	insinto /usr/"$(get_libdir)"/pkgconfig
+	sed "s|@LIBDIR@|$(get_libdir)|;s|@VERSION@|${PV}|;" "${FILESDIR}/libyuv.pc" |
```
Hmmm, should the live ebuild set the version to 9999 here? That seems wrong. Grep it out of `include/libyuv/version.h` instead?

eli-schwartz replied:
Upstream doesn't have a pkgconfig file at all. Neither does Debian. Why should we add one?

If one did get added it should be configured by cmake rather than via a sed in src_install. That is (again) best done upstream.

N-R-K replied:
> Neither does Debian.

Huh, I thought they did. Fedora does though.

> Why should we add one?

If not then patching imlib2 would be needed.

> That is (again) best done upstream.

Agreed. But I'm not interested in making a google account to sign CLA in order to do so.

eli-schwartz replied:
> If not then patching imlib2 would be needed.
> 
> > That is (again) best done upstream.
> 
> Agreed. But I'm not interested in making a google account to sign CLA in order to do so.

You could instead submit an upstream patch to imlib2, which would make imlib2 build with libyuv support on distros such as Debian which don't have that pkg-config file but do have libyuv. No CLA required in that case.

N-R-K replied:
> You could instead submit an upstream patch to imlib2, which would make imlib2 build with libyuv support on distros such as Debian which don't have that pkg-config file but do have libyuv. No CLA required in that case.

Not super enthusiastic about patching autoconf stuff :)

I opened an issue letting imlib2 know about it, though: https://git.enlightenment.org/old/legacy-imlib2/issues/31

eli-schwartz replied:
@xen0n I wonder, since I happened to notice you have a Google gerrit accound were recently submitting patches for libyuv, would you be interested in e.g. contributing an upstream pkg-config file, fixing the install directory, etc?

N-R-K commented:
Okay, I think I'm mostly happy with the current state. Rebased on top of master to resolve some `package.use.mask` conflicts.

eli-schwartz reviewed (approved):
Thanks for the updates w.r.t. deferring imlib2[y4m] until the pkg-config issue is solved.

Have not tested the PR but based on eyeball review this lgtm.

dlan17 commented:
> Thanks for the updates w.r.t. deferring imlib2[y4m] until the pkg-config issue is solved.
> 
> Have not tested the PR but based on eyeball review this lgtm.

thanks for the review, I will give it a test and then proceed..

dlan17 commented on media-libs/libavif/libavif-1.2.0.ebuild:
```diff
@@ -0,0 +1,113 @@
+# See bug #822336 re subslot
+SLOT="0/16.1.1"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~loong ~mips ~ppc64 ~riscv ~sparc ~x86"
+IUSE="+aom dav1d examples extras gdk-pixbuf rav1e svt-av1 libyuv test"
```
how about just naming the USE as 'yuv'? it's more intuitive which introduce YUV scaling & conversion support..

N-R-K replied:
> it's more intuitive which introduce YUV scaling & conversion support..

`libavif` already has yuv support, `libyuv` doesn't introduce new functionality, just speeds them up: https://github.com/AOMediaCodec/libavif/?tab=readme-ov-file#libyuv-dependency

dlan17 replied:
> `libavif` already has yuv support, `libyuv` doesn't introduce new functionality, just speeds them up

Ok, then be more libyuv specific, I'm fine, thanks

dlan17 commented:
got one QA failure, but when I checked, it seems not introduced in this PR..

media-libs/libavif
  UnstableOnly: for arch: [ hppa ], all versions are unstable: [ 1.1.1, 1.2.0 ]
  PotentialStable: version 1.1.1: slot(0), stabled arches: [ amd64, arm, arm64, ppc64, sparc, x86 ], potential: [ ~hppa ]
  RequiredUseDefaults: version 1.1.1: profile: 'default/linux/mips/23.0/mipsel/multilib/n32' (19 total) failed REQUIRED_USE: ( aom || dav1d )
  RequiredUseDefaults: version 1.2.0: profile: 'default/linux/mips/23.0/mipsel/multilib/n32' (19 total) failed REQUIRED_USE: ( aom || dav1d )

N-R-K commented:
> got one QA failure, but when I checked, it seems not introduced in this PR..

Yes, I had this too. But it was already present in older version 1.1.1 and so I wasn't sure what to do there.

eli-schwartz commented:
```
commit fa7a014ae15351103f9c5388afc661c06cd69977 (HEAD, origin/master, origin/HEAD)
Author:     Eli Schwartz <eschwartz@gentoo.org>
AuthorDate: Tue Mar 18 22:50:34 2025
Commit:     Eli Schwartz <eschwartz@gentoo.org>
CommitDate: Tue Mar 18 22:52:40 2025

    profiles/arch/mips: make libavif installable by default
    
    Fixes pkgcheck failure for RequiredUseDefaults. Mimics the identical fix
    already applied in profiles/features/big-endian.
    
    Signed-off-by: Eli Schwartz <eschwartz@gentoo.org>

 profiles/{features/big-endian => arch/mips}/package.use | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)
 copy profiles/{features/big-endian => arch/mips}/package.use (65%)
```

Fixed.

novomesk commented:
media-libs/libavif-1.2.0::gentoo is problematic. It doesn't build on my system:

```
-- Found ZLIB: /usr/lib64/libz.so (found version "1.3.1")
-- Found PNG: /usr/lib64/libpng.so (found suitable version "1.6.44+apng", minimum required is "1.6.32")
-- Found JPEG: /usr/lib64/libjpeg.so (found version "62")
-- libavif: libxml2 not found; avifenc will ignore any gain map in JPEG files
-- libavif(AVIF_LIBARGPARSE=LOCAL): compiled library not found at /var/tmp/portage/media-libs/libavif-1.2.0/w         ork/libavif-1.2.0/ext/libargparse/build/libargparse.a; using FetchContent
-- libavif(AVIF_LIBARGPARSE=LOCAL): fetching and configuring libargparse
[1/9] Creating directories for 'libargparse-populate'
[1/9] Performing download step (git clone) for 'libargparse-populate'
Cloning into 'libargparse-src'...
fatal: unable to access 'https://github.com/kmurray/libargparse.git/': Could not resolve host: github.com
Cloning into 'libargparse-src'...
fatal: unable to access 'https://github.com/kmurray/libargparse.git/': Could not resolve host: github.com
Cloning into 'libargparse-src'...
fatal: unable to access 'https://github.com/kmurray/libargparse.git/': Could not resolve host: github.com
Had to git clone more than once: 3 times.
CMake Error at libargparse-subbuild/libargparse-populate-prefix/tmp/libargparse-populate-gitclone.cmake:50 (m         essage):
  Failed to clone repository: 'https://github.com/kmurray/libargparse.git'
```
Other have the similar problem too: https://bugs.gentoo.org/951614

There is also report about unused variable:
https://bugs.gentoo.org/951618

N-R-K commented:
> https://bugs.gentoo.org/951614

I had `extras` disabled so did not notice this. Apologies. Should be fixed by https://github.com/gentoo/gentoo/pull/41205.
